libursa/src/bn/rust.rs

```diff
@@ -0,0 +1,632 @@
+        Ok(BigNumber { bn: res })
+    }
+
+    pub fn add_word(&mut self, w: u32) -> Result<&mut BigNumber, UrsaCryptoError> {
+        self.bn += w;
+        Ok(self)
+    }
+
+    pub fn sub_word(&mut self, w: u32) -> Result<&mut BigNumber, UrsaCryptoError> {
+        self.bn -= w;
+        Ok(self)
+    }
+
+    pub fn mul_word(&mut self, w: u32) -> Result<&mut BigNumber, UrsaCryptoError> {
+        self.bn *= w;
+        Ok(self)
+    }
+
+    pub fn div_word(&mut self, w: u32) -> Result<&mut BigNumber, UrsaCryptoError> {
+        self.bn /= w;
+        Ok(self)
+    }
+
+    pub fn mod_exp(&self, a: &BigNumber, b: &BigNumber, _ctx: Option<&mut BigNumberContext>) -> Result<BigNumber, UrsaCryptoError> {
+        if a.is_negative() {
+            let res = self.inverse(&b, _ctx)?;
+            let a = a.set_negative(false)?;
+            Ok(BigNumber{ bn: res.bn.modpow(&a.bn, &b.bn) })
+        } else {
+            let res = self.bn.modpow(&a.bn, &b.bn);
+            Ok(BigNumber { bn: res })
+        }
+    }
+
+    pub fn modulus(&self, a: &BigNumber, _ctx: Option<&mut BigNumberContext>) -> Result<BigNumber, UrsaCryptoError> {
+        let res = &self.bn % &a.bn;
+        Ok(BigNumber { bn: res })
+    }
+
+    pub fn exp(&self, a: &BigNumber, _ctx: Option<&mut BigNumberContext>) -> Result<BigNumber, UrsaCryptoError> {
+        if self.bn.bits() == 0 {
+            return Ok(BigNumber::default())
+        } else if a.bn.is_one() {
+            return Ok(self.clone()?)
+        }
+
+        match a.bn.to_u64() {
+            Some(num) => Ok(BigNumber { bn: self.bn.pow(num) }),
+            None => Err(UrsaCryptoError::InvalidStructure("'a' cannot be help in u64".to_string()))
+        }
+    }
+
+    pub fn inverse(&self, n: &BigNumber, _ctx: Option<&mut BigNumberContext>) -> Result<BigNumber, UrsaCryptoError> {
+        if n.bn.is_one() ||
+           n.bn.is_zero() {
+            return Err(UrsaCryptoError::InvalidStructure("Invalid modulus".to_string()))
+        }
+
+        // Euclid's extended algorithm, Bèzout coefficient of `n` is not needed
+        //n is either prime or coprime
+        //
+        //function inverse(a, n)
+        //    t := 0;     newt := 1;
+        //    r := n;     newr := a;
+        //    while newr ≠ 0
+        //        quotient := r div newr
+        //        (t, newt) := (newt, t - quotient * newt)
+        //        (r, newr) := (newr, r - quotient * newr)
+        //    if r > 1 then return "a is not invertible"
+        //    if t < 0 then t := t + n
+        //    return t
+        //
+        let (mut t, mut new_t) = (BigInt::zero(), BigInt::one());
+        let (mut r, mut new_r) = (n.bn.clone(), self.bn.clone());
+
+        while !new_r.is_zero() {
+            let quotient = &r / &new_r;
+            let temp_t = t.clone();
+            let temp_new_t = new_t.clone();
+
+            t = temp_new_t.clone();
+            new_t = temp_t - &quotient * temp_new_t;
```

Comment:
Moving `new_t = temp_t - &quotient * temp_new_t;` above `t = temp_new_t.clone();` should save the clone of `t`, i.e `temp_t`, right?